ark/phenotyping/cluster_helpers.py

```diff
@@ -1,16 +1,309 @@
+            warnings.warn('Pixel SOM already trained')
+            return
+
+        super().train_som(self.train_data[self.columns])
+
+    def assign_som_clusters(self, external_data: pd.DataFrame) -> pd.DataFrame:
+        """Assigns SOM clusters using `weights` to a dataset
+
+        Args:
+            external_data (pandas.DataFrame):
+                The dataset to assign SOM clusters to
+
+        Returns:
+            pandas.DataFrame:
+                The dataset with the SOM clusters assigned.
+        """
+        # normalize external_data prior to assignment
+        external_data_norm = self.normalize_data(external_data)
+        som_labels = super().generate_som_clusters(external_data_norm)
+
+        # assign SOM clusters to external_data
+        external_data_norm['pixel_som_cluster'] = som_labels
+
+        return external_data_norm
+
+
+class CellSOMCluster(PixieSOMCluster):
+    def __init__(self, cell_data_path: pathlib.Path, weights_path: pathlib.Path,
+                 columns: List[str], num_passes: int = 1, xdim: int = 10, ydim: int = 10,
+                 lr_start: float = 0.05, lr_end: float = 0.01):
+        """Creates a cell SOM cluster object derived from the abstract PixieSOMCluster
+
+        Args:
+            cell_data_path (pathlib.Path):
+                The name of the cell dataset to use for training
+            weights_path (pathlib.Path):
+                The path to save the weights to.
+            columns (List[str]):
+                The list of columns to subset the data on.
+            num_passes (int):
+                The number of SOM training passes to use.
+            xdim (int):
+                The number of SOM nodes on the x-axis.
+            ydim (int):
+                The number of SOM nodes on the y-axis.
+            lr_start (float):
+                The initial learning rate.
+            lr_end (float):
+                The learning rate to decay to
+        """
+        super().__init__(
+            weights_path, columns, num_passes, xdim, ydim, lr_start, lr_end
+        )
+
+        # path validation
+        validate_paths([cell_data_path])
+        self.cell_data_path = cell_data_path
+
+        # load the cell data in
+        self.cell_data = feather.read_dataframe(cell_data_path)
+
+        # since cell_data is the only dataset, we can just normalize it immediately
+        self.cell_data = self.normalize_data()
+
+    def normalize_data(self) -> pd.DataFrame:
+        """Uses `norm_data` to normalize `cell_data`
+
+        Returns:
+            pandas.DataFrame:
+                `cell_data` with `columns` normalized by the values in `norm_data`
+        """
+        # only 99.9% normalize on the columns provided
+        cell_data_sub = self.cell_data[self.columns].copy()
+        cell_data_sub = cell_data_sub.div(cell_data_sub.quantile(0.999))
+
+        # assign back to cell_data
+        self.cell_data[self.columns] = cell_data_sub
```

Comment:
My python oos might just be a little rusty - is it necessary to assign `self.cell_data` here, and then return it, and then assign `self.cell_data` in the init function above (line 258)? To mirror the `normalize_data` function in the pixel version, maybe we can just return cell_data_sub here and assign it to `self.cell_data` in the init function.

Reply:
Good catch. I was hoping to avoid having ugly type hints for the checker but think I found a way to easily resolve that.